content/client-lib-development-guide/features.textile

```diff
@@ -128,7 +141,19 @@ h3(#rest-channel). Channel
+** @(RSL1e)@ Allows @name@ and or @data@ to be @null@.  If any of the values are null, then key is not sent to Ably i.e. a payload with a @null@ value for @data@ would be sent as follows @{ "name": "click" }@
+** @(RSL1f)@ Unauthenticated clients without a @clientId@ (any @clientId@ is permitted):
+*** @(RSL1f1)@ When a @Message@ with a @clientId@ value is published, Ably will accept and publish that message with the provided @clientId@. A test should assert via the history API that the @clientId@ of the published @Message@ is correct for both authentication schemes (token and basic)
+** @(RSL1g)@ Authenticated clients with a @clientId@ (as a result of either an explicitly configured @clientId@ in @ClientOptions@, or implicitly through token auth):
+*** @(RSL1g1)@ When publishing a @Message@ with the @clientId@ attribute set to @null@:
+**** @(RSL1g1a)@ It is unnecessary for the client to set the @clientId@ of the @Message@ before publishing
+**** @(RSL1g1b)@ Ably is expected to automatically assign a @clientId@ upon receiving the @Message@. A test should assert via the history API that the @clientId@ value is populated for the @Message@ when received
+*** @(RSL1g2)@ When publishing a @Message@ with the @clientId@ attribute value set to the authenticated client's @clientId@, Ably will accept the message and publish it. A test should assert that the @clientId@ value is populated for the @Message@ when received
+*** @(RSL1g3)@ When publishing a @Message@ with a different @clientId@ attribute value to the authenticated @clientId@, the client library should reject the message, and an exception should be raised. The connection and channel remains available for further operations
+*** @(RSL1g4)@ When using the Realtime library, it is possible to publish a message with an explicit @clientId@ before the library has completed authentication. As such, it is possible that the @clientId@ inferred following authentication is incompatible with the explicit @clientId@ of the message queued for delivery. In this case, it is not expected that the client library will reject the already queued message as the Ably server will reject the message anyway when received. However, all future messages published with an incompatible @clientId@ should be immediately rejected by the client library
+*** @(RSL1g5)@ When using the REST library, publishing a message with an explicit @clientId@ that is incompatible with the @clientId@ inferred following authentication will cause the library to reject the message immediately. Unlike the Realtime library, the message is never queued,and as such, authentication will be completed synchronously before attempting to publish the message over REST
+** Where the library language permits, the @Channel#publish(name, data)@ method should provide an optional argument that allows the @clientId@ value to be specified such as @Channel#publish('event', 'data', { clientId: 'John' })@
```

Comment:
Thanks, it was caused by a merge issue.  The branch when merged did not have the codes in it, so was a nightmare.  Thanks for spotting.